core/commands/corenet.go

```diff
@@ -0,0 +1,566 @@
+	ma "gx/ipfs/QmcyqRMCAXVtYPS4DiBrA7sezL9rRGfW8Ctx7cywL4TXJj/go-multiaddr"
+	peer "gx/ipfs/QmdS9KpbDyPrieswibZhkod1oXqRwZJrUPzxCofAMWpFGq/go-libp2p-peer"
+	manet "gx/ipfs/Qmf1Gq7N45Rpuw7ev47uWgH6dLPtdnvcMRNPkVBwqjLJg2/go-multiaddr-net"
+)
+
+// Command output types.
+type AppInfoOutput struct {
+	Protocol string
+	Address  string
+}
+
+type StreamInfoOutput struct {
+	HandlerId     string
+	Protocol      string
+	LocalPeer     string
+	LocalAddress  string
+	RemotePeer    string
+	RemoteAddress string
+}
+
+type CorenetLsOutput struct {
+	Apps []AppInfoOutput
+}
+
+type CorenetStreamsOutput struct {
+	Streams []StreamInfoOutput
+}
+
+// cnAppInfo holds information on a local application protocol listener service.
+type cnAppInfo struct {
+	// Application protocol identifier.
+	protocol string
+
+	// Node identity
+	identity peer.ID
+
+	// Local protocol stream address.
+	address ma.Multiaddr
+
+	// Local protocol stream listener.
+	closer io.Closer
+
+	// Flag indicating whether we're still accepting incoming connections, or
+	// whether this application listener has been shutdown.
+	running bool
+}
+
+func (c *cnAppInfo) Close() error {
+	apps.Deregister(c.protocol)
+	c.closer.Close()
+	return nil
+}
+
+// cnAppRegistry is a collection of local application protocol listeners.
+type cnAppRegistry struct {
+	apps []*cnAppInfo
+}
+
+func (c *cnAppRegistry) Register(appInfo *cnAppInfo) {
+	c.apps = append(c.apps, appInfo)
+}
+
+func (c *cnAppRegistry) Deregister(proto string) {
+	foundAt := -1
+	for i, a := range c.apps {
+		if a.protocol == proto {
+			foundAt = i
+			break
+		}
+	}
+
+	if foundAt != -1 {
+		c.apps = append(c.apps[:foundAt], c.apps[foundAt+1:]...)
+	}
+}
+
+// cnStreamInfo holds information on active incoming and outgoing protocol app streams.
+type cnStreamInfo struct {
+	handlerId uint64
+
+	protocol string
+
+	localPeer peer.ID
+	localAddr ma.Multiaddr
+
+	remotePeer peer.ID
+	remoteAddr ma.Multiaddr
+
+	local  io.ReadWriteCloser
+	remote io.ReadWriteCloser
+}
+
+func (c *cnStreamInfo) Close() error {
+	c.local.Close()
+	c.remote.Close()
+	streams.Deregister(c.handlerId)
+	return nil
+}
+
+// cnStreamRegistry is a collection of active incoming and outgoing protocol app streams.
+type cnStreamRegistry struct {
+	streams []*cnStreamInfo
+
+	nextId uint64
+}
+
+func (c *cnStreamRegistry) Register(streamInfo *cnStreamInfo) {
+	streamInfo.handlerId = c.nextId
+	c.streams = append(c.streams, streamInfo)
+	c.nextId += 1
+}
+
+func (c *cnStreamRegistry) Deregister(handlerId uint64) {
+	foundAt := -1
+	for i, s := range c.streams {
+		if s.handlerId == handlerId {
+			foundAt = i
+			break
+		}
+	}
+
+	if foundAt != -1 {
+		c.streams = append(c.streams[:foundAt], c.streams[foundAt+1:]...)
+	}
+}
+
+//TODO: Ideally I'd like to see these combined into a module in core.
+var apps cnAppRegistry
+var streams cnStreamRegistry
+
+var CorenetCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Application network streams.",
```

Comment:
tagline: "libp2p stream mounting"
shortdescription: "expose a local application to remote peers over libp2p

Note: this command is experimental and subject to change as usecases and APIs are refined"

cc @Kubuxu @lgierth